### PR TITLE
MRG: improve restart by optionally writing batched zipfiles

### PR DESCRIPTION
**Pull request by bluegenes**
This PR introduces a new optional param `--batch-size`, which allows users to build smaller zipfiles with `gbsketch` or `urlsketch`. These zipfiles are populated sequentially, with all signatures associated with `batch_size` accessions (not `batch_size` signatures). If `gbsketch`/`urlsketch` fail, they can read any zipfiles that were finished in order to restart. Zip names will be generated from the `--output`, so if output is `output.zip`, batches will be `output.1.zip`, `output.2.zip`, etc.  I'm not really sure what `batch_size` to recommend, but I think the overhead is fairly low for creating new small zips -- the main issue will be if users later want to concatenate them into a single zip.

Uses the changes from #101 to enable writing batched zipfiles as a way to improve restart.

- [x] make batch_size a user modifiable parameter
- [x] For cases where the total number of signatures is less than the `batch_size`, we could write the regular `*zip` file, with no `.1`, etc.
- [x] functions to enable reading from existing batched zips to allow restart
  - [x] build filename: paramset Hashmap, use that to filter the template sigs for each filename using `filter`
- [x] add tests for batched zipfile writing, recovery from existing batches
- [x] move zip_writer creation inside writing loop to avoid empty final zip
- [x] check what happens if we have an unclosed zip (i.e. from unexpected failure)
  - **sourmash panics on invalid zips. Here I've caught the panic and ignored it**, but it may ultimately be better to handle + return error at the sourmash level (`ZipStorage::from_file` panics)
  - Note that we will likely have an invalid zip upon any restart from failure, because the zip file would not have properly been closed/finished.

Issue for later: 
- #107 

Fixes:
- #69

### CODE REVIEW

**Comment by bluegenes** [comment]
@ctb ready for review

@ccbaumler - better later than never, right? Does this fit your use cases? Do you have a recommendation for the # of files we should recommend in each batch?

File: README.md [hunk 1]
```diff
@@ -156,9 +172,19 @@ To run the test accession file at `tests/test-data/acc-url.csv`, run:
 sourmash scripts urlsketch tests/test-data/acc-url.csv -o test-urlsketch.zip -f out_fastas -k --failed test.failed.csv -p dna,k=21,k=31,scaled=1000,abund -p protein,k=10,scaled=100,abund -r 1
 ```
 
+### Usage Considerations
```
Comment (ctb) [inline README.md:175]: duplicated text?

File: src/utils.rs [hunk 2]
```diff
@@ -844,3 +869,337 @@ pub fn parse_params_str(params_strs: String) -> Result<Vec<Params>, String> {
 
     Ok(unique_params.into_iter().collect())
 }
+
+// this should be replaced with branchwater's MultiCollection when it's ready
```
Comment (ctb) [inline src/utils.rs:873]: maybe create an issue in branchwater repo? I didn't really design around re-use 😅 - you could also link to https://github.com/sourmash-bio/sourmash/issues/3321.

so, ok, suggest:
- create issue in this repo
- link to sourmash issue https://github.com/sourmash-bio/sourmash/issues/3321

File: README.md (hunk 1 above)
Comment (bluegenes) [reply README.md:175]: one for each command. But i can bump up a level and just include it once.

File: src/utils.rs (hunk 2 above)
Comment (bluegenes) [reply src/utils.rs:873]: sure, will do. Note, it doesn't really need to be replaced with `MultiCollection` - just, if we eventually get `MultiCollection` in sourmash core, it'd be good to just have one supported implementation.

File: README.md (hunk 1 above)
Comment (bluegenes) [reply README.md]: done in [b583fb9](https://github.com/sourmash-bio/sourmash_plugin_directsketch/pull/102/commits/b583fb9a04c6df61b0ae7903fd2a12adb7c9bf7f)

**Pull request merged**